crates/extensions/tedge_flows/src/js_lib/kv_store.rs

```diff
@@ -0,0 +1,115 @@
+use crate::js_value::JsonValue;
+use rquickjs::class::Trace;
+use rquickjs::Ctx;
+use rquickjs::IntoJs;
+use rquickjs::JsLifetime;
+use rquickjs::Object;
+use rquickjs::Result;
+use rquickjs::Value;
+use std::collections::BTreeMap;
+use std::collections::HashMap;
+use std::ops::Deref;
+use std::sync::Arc;
+use std::sync::Mutex;
+
+#[derive(Clone, Default, JsLifetime)]
+pub struct KVStore {
+    data: Arc<Mutex<HashMap<String, BTreeMap<String, JsonValue>>>>,
+}
+
+pub const MAPPER_NAMESPACE: &str = "";
+
+impl KVStore {
+    pub fn init(&self, ctx: &Ctx<'_>) {
+        self.store_as_userdata(ctx)
+    }
+
+    pub fn js_context<'js>(
+        ctx: &Ctx<'js>,
+        flow_name: &str,
+        script_name: &str,
+        config: &JsonValue,
+    ) -> Result<Value<'js>> {
+        let context = Object::new(ctx.clone())?;
+
+        context.set("mapper", FlowStore::new(MAPPER_NAMESPACE))?;
+        context.set("flow", FlowStore::new(flow_name))?;
+        context.set("script", FlowStore::new(script_name))?;
```

Comment:
What's the use-case you had in mind for `script` context? From a user POV, how is it different from a `static` state variable defined in a script?

I'm asking this based on the following interpretation of the context hierarchy:
* `mapper`: Shared across all flows
* `flow`: Shared between steps
* `script`: Shared with what?

Reply:
> What's the use-case you had in mind for `script` context? From a user POV, how is it different from a `static` state variable defined in a script?

The plan is to persist the script context across restarts. Not done yet. Currently the state is only kept in memory and persists module reload.

> I'm asking this based on the following interpretation of the context hierarchy:
> * `mapper`: Shared across all flows
> * `flow`: Shared between steps
> * `script`: Shared with what?

* `script`: Persisted across reloads and restarts